### PR TITLE
Disable kubelet read-only port by default (only set the correct default value)

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind deprecation

Jordon did an accurate summary of it in https://github.com/kubernetes/kubernetes/pull/100335#issuecomment-1397078159 like below.

> kubelet has different defaults when starting from a config file and from command-line flags... CLI flags came first and defaulted the readonly port to 10250. kubelet file-based config came next and we used that opportunity to improve defaults (see #59666)
> 
> I still think this is a reasonable step, but it is not a bug.
> 
> All the test changes need review/approve from kubelet owners to make sure the surface area is still tested

#### What this PR does / why we need it:

As the `kubelet --help` shows, the default value is 10255, the default value is 0 indeed. (If you don't specify the ` --read-only-port ` in 1.21 the port is disabled)
> --read-only-port int32    The read-only port for the Kubelet to serve on with no authentication/authorization (set to 0 to disable) (default 10255) (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)

**To clarify, the help docs above are not all correct.**
- 1.21:   (default 10255) ❌  【this is a doc issue】
- 1.21:   The default value of kubelet read-only port is now 0 if not set. ✅
- 1.21：Set to 0 to disable. ✅
- 1.29 with this fix： The default value of kubelet read-only port is now 0 if not set. ✅
- 1.29 with this fix：Set to 0 to disable. ✅
- 1.29 with this fix： --read-only-port=10255 to reenable it.
-  To be determined: 1.30?+: --read-only-port will be removed (as @tallclair  mentioned in https://github.com/kubernetes/kubernetes/issues/12968#issuecomment-622613486, it should not be removed after we fix those problems that are raised in the comment)

**This PR just changes the default behavior to what it says it would.**

**What did I do in this PR?**

- [x] correct the default value of flag `--read-only-port`
- [x] remove the default port 10255 as well as KubeletReadOnlyPort
- [x] use `:10250/stats/summary` instead of `:10255/stats/summary` in e2e test
- [x] use `:10248/healthz` instead of `:10255/healthz` in tests
- [ ]  use `:10250/metrics` instead of `:10255/metrics` in e2e test

#### Which issue(s) this PR fixes:
xrefs #12968

#### Special notes for your reviewer:
BTW, we should deprecate this configure later( we may discuss it in  #12968 )

#### Does this PR introduce a user-facing change?

```release-note
ACTION REQUIRED: The kubelet read-only port is disabled by default. Set `readOnlyPort` to 10255 in kubelet's config file to reenable it.
```

### CODE REVIEW

**Review by troy0820 — COMMENTED** [review]
/lgtm

**Comment by pacoxu** [comment]
/retest
Before Suite flake

**Comment by derekwaynecarr** [comment]
I am also seeing the default readonly is now 0, but was a little confused figuring out how that was the case given the config file and flag defaulting that happens and had to refresh my own personal memory.  From what I see the config file defaults to 0, and then the flag set chooses its default from that file when adding the kubelet config flags.  From that perspective, the change here makes sense.

/test pull-kubernetes-node-e2e

**Comment by pacoxu** [comment]
/assign @lavalamp 
for approval

/retest
for flake 😓

**Comment by lavalamp** [comment]
I am not a kubelet expert but setting it to 0 means you get an unpredictable ephemeral port in practice, right? If that is the case I think it is drastically safer to disable it than default it to 0. I think if the default accidentally changed to zero due to reading from the config file, we need to fix that *and backport the fix*. Having a random read-only port sounds like an ingredient in a security vuln (but not a vuln by itself).

**Comment by tallclair** [comment]
I'm fine with changing the default, but I don't think the concerns raised in https://github.com/kubernetes/kubernetes/issues/12968#issuecomment-622613486 have been adequately addressed to remove support for this entirely.

> I am not a kubelet expert but setting it to 0 means you get an unpredictable ephemeral port in practice, right? If that is the case I think it is drastically safer to disable it than default it to 0.

I agree that it's a scary mechanism, but we explicitly read 0 as disabled: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/cmd/kubelet/app/server.go#L1203-L1205

**Comment by tallclair** [comment]
This also needs a release note.

**Comment by pacoxu** [comment]
Release note updated. 

The default 0 came from `KubeletConfiguration` default settings. It would override the port if flag `--read-only-port` is not set.

/retest

**Comment by pacoxu** [comment]
/test pull-kubernetes-node-e2e

https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/794/files fixed the problem hours ago.

**Comment by pacoxu** [comment]
> I0415 05:22:36.762] W0415 05:20:39.921024     984 util.go:104] Health check on "http://127.0.0.1:10255/healthz" failed, error=Head "http://127.0.0.1:10255/healthz": dial tcp 127.0.0.1:10255: connect: connection refused

Fix the e2e failure as it use the deprecated port to check kubelet health.

**Comment by pacoxu** [comment]
> I0415 05:22:36.762] W0415 05:20:39.921024     984 util.go:104] Health check on "http://127.0.0.1:10255/healthz" failed, error=Head "http://127.0.0.1:10255/healthz": dial tcp 127.0.0.1:10255: connect: connection refused

Fix the e2e failure as it uses the deprecated port to check kubelet health.

File: pkg/cluster/ports/ports.go
Comment (ehashman) [inline pkg/cluster/ports/ports.go:34]: The comment was removed but not the port constant?

**Review by ehashman — CHANGES_REQUESTED** [review]
I'd update your release note to be a bit more clear. I think this needs an action required:

```
ACTION REQUIRED: The kubelet read-only port is disabled by default and will be removed in a future release. Set the kubelet flag `--read-only-port=10255` to reenable it.
```

File: pkg/cluster/ports/ports.go
Comment (pacoxu) [reply pkg/cluster/ports/ports.go:34]: The port is still used in kubemark.

We may remove this port once read-only port is not supported.

**Comment by pacoxu** [comment]
/skip
for the removed test pull-kubernetes-node-e2e

File: pkg/cluster/ports/ports.go
Comment (pacoxu) [reply pkg/cluster/ports/ports.go]: I have updated the e2e test.

**Comment by pacoxu** [comment]
Updated. I think this PR is ready to review now. @tallclair @derekwaynecarr @lavalamp @ehashman. Would you take a look again?

1. correct the default value of flag --read-only-port
2. remove the default port 10255 as well as KubeletReadOnlyPort
3. use :10250/stats/summary instead of :10255/stats/summary` in e2e test
4. use :10248/healthz instead of :10255/healthz in tests

**Comment by dims** [comment]
@SergeyKanzhelev did you get a chance to review this?

**Comment by pacoxu** [comment]
> @SergeyKanzhelev did you get a chance to review this?

rebased.

**Comment by Nirusu** [comment]
@pacoxu How are you supposed to understand this roadmap? Is the read-only port already supposed to be disabled by default since v1.21 and this is an issue with the documentation? Or is 10255 still the default as of v1.23, and the plan is to disable the read-only port by default for a _future_ release (and then even further in the future, remove it completely)?

**Comment by pacoxu** [comment]
> @pacoxu How are you supposed to understand this roadmap? Is the read-only port already supposed to be disabled by default since v1.21 and this is an issue with the documentation? Or is 10255 still the default as of v1.23, and the plan is to disable the read-only port by default for a _future_ release (and then even further in the future, remove it completely)?

Firstly the problem is an issue with the documentation. The default value of the flag is 0, not 10255. (You can try to test. The cause is `--config-file` default will take effect if `--ready-only-port` is not set).

Another opinion is that we should not use 10255 since it is by default disabled.

For future releases, I cannot decide it but we could remove it if we want since it is disabled by default and there are alternative ways for it.

**Comment by pacoxu** [comment]
@SergeyKanzhelev do you have time to review this agian?

**Comment by pacoxu** [comment]
/remove-lifecycle stale
wait for review

**Comment by neoaggelos** [comment]
Hi @dims, 1.26 bug triage team here, just wanted to check whether this Is this still on track for the 1.26 milestone.

Thank you!

File: cmd/kubemark/hollow-node.go
Comment (SergeyKanzhelev) [inline cmd/kubemark/hollow-node.go:94]: why is it being removed here?

**Comment by SergeyKanzhelev** [comment]
> yep, i am hoping sig-node reviewers/approvers will consider this for 1.26

I thought I mentioned before, test changes if sent as separate PR can be merged any moment. Moving away from using this port in tests is not a problem.

Changing default to disabled sounds ok as well just to make installations more secure.

I do not agree with the message in the release notes. Yes, we have plans to offer an alternative for readonlyport. But so far we do not have a clear path for people to migrate. This is very discouraging to announce deprecation of something without offering an alternative for critical user scenarios.

File: cmd/kubemark/hollow-node.go
Comment (pacoxu) [reply cmd/kubemark/hollow-node.go:95]: we may change the default value to kubelet default 0 here and keep the flag.

**Comment by pacoxu** [comment]
> This is very discouraging to announce deprecation of something without offering an alternative for critical user scenarios.

So it means that we will remove it once we have an alternative way.

As this PR only changes the default behavior to what it says it would be. I think we can change the release note to not mention the deprecation or add the `once there is an alternative` in it. I prefer the second, however the first one below is more simple to understand.


```
ACTION REQUIRED: The kubelet read-only port is disabled by default. Set the kubelet flag `--read-only-port=10255` to reenable it.
```

or

```
ACTION REQUIRED: The kubelet read-only port is disabled by default and will be removed in a future release once there is an alternative. Set the kubelet flag `--read-only-port=10255` to reenable it.
```

File: test/e2e/cloud/gcp/kubelet_security.go
Comment (pacoxu) [reply test/e2e/cloud/gcp/kubelet_security.go]: Updated.

Use `LegacyKubeletReadOnlyPort` for e2e test cases.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

File: cmd/kubemark/hollow-node.go
Comment (SergeyKanzhelev) [inline cmd/kubemark/hollow-node.go:95]:
```suggestion
	fs.IntVar(&c.KubeletReadOnlyPort, "kubelet-read-only-port", 0, "Read-only port on which Kubelet is listening. The value `0` indicates that the port is disabled.")
```

this may be a valuable addition as some systems uses `0` to pick a random

File: pkg/cluster/ports/ports.go
Comment (SergeyKanzhelev) [inline pkg/cluster/ports/ports.go:29]: let's move to test where it is used.

**Review by SergeyKanzhelev — APPROVED** [review]
two minor comments, lgtm otherwise

Thank you!

File: cmd/kubemark/hollow-node.go
Comment (pacoxu) [reply cmd/kubemark/hollow-node.go]: Updated.

File: pkg/cluster/ports/ports.go
Comment (pacoxu) [reply pkg/cluster/ports/ports.go]: Move to `test/e2e/network/firewall.go` where it is used.

File: test/e2e/cloud/gcp/kubelet_security.go
Comment (SergeyKanzhelev) [inline test/e2e/cloud/gcp/kubelet_security.go:51]: fmt.Sprintf is not needed since you removed the substitute

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

File: test/e2e/cloud/gcp/kubelet_security.go
Comment (pacoxu) [reply test/e2e/cloud/gcp/kubelet_security.go]: updated.

**Comment by sftim** [comment]
The proposed changelog entry suggests setting a deprecated command line argument. Should we recommend a config file change instead?

**Comment by pacoxu** [comment]
> The proposed changelog entry suggests setting a deprecated command line argument. Should we recommend a config file change instead?

Good point. Updated.

/assign @smarterclayton @liggitt

File: cmd/kube-apiserver/app/options/options_test.go
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options_test.go:111]: should this file get reverted? it's testing that the flag is functional, which it still should be, right?

File: pkg/cluster/ports/ports.go
Comment (liggitt) [inline pkg/cluster/ports/ports.go:35]: revert the change to KubeletHealthzPort godoc

File: test/e2e_node/eviction_test.go
Comment (liggitt) [inline test/e2e_node/eviction_test.go:357]: the changes to this file don't seem related to the readonly port change, should this file be reverted?

File: test/e2e_node/util.go
Comment (liggitt) [inline test/e2e_node/util.go:280]: does the healthz port serve metrics??

File: test/e2e_node/util.go
Comment (liggitt) [inline test/e2e_node/util.go:94]: can we not provide a credential so the non-readonly port can be used? how many e2e tests is this skipping?

File: test/e2e/storage/vsphere/vsphere_volume_master_restart.go
Comment (liggitt) [inline test/e2e/storage/vsphere/vsphere_volume_master_restart.go:45]: is this being exercised in presubmit?

Comment (pacoxu) [reply test/e2e/storage/vsphere/vsphere_volume_master_restart.go:45]: This may not be exercised in presubmit as `[Feature:vsphere][Serial][Disruptive]`.

Either port should work. Since read-only port will be disabled by default, alter it to use the health port.
```
[root@paco-centos-9 ~]# curl http://127.0.0.1:10248/healthz
ok[root@paco-centos-9 ~]# curl http://127.0.0.1:10255/healthz
ok
```

File: test/e2e_node/eviction_test.go
Comment (pacoxu) [reply test/e2e_node/eviction_test.go:358]: reverted.

File: cmd/kube-apiserver/app/options/options_test.go
Comment (pacoxu) [reply cmd/kube-apiserver/app/options/options_test.go]: we can keep this flag ut.

File: pkg/cluster/ports/ports.go
Comment (pacoxu) [reply pkg/cluster/ports/ports.go]: done.

File: test/e2e_node/util.go
Comment (pacoxu) [reply test/e2e_node/util.go:280]: My fault.

For metrics, it is similar to `/stats/summary`. 
- https + 10250 
- or HTTP + 10255

I kept this and marked `GrabKubeletMetricsWithoutProxy` as deprecated. We should alter to use `10250`.

File: test/e2e_node/util.go
Comment (pacoxu) [reply test/e2e_node/util.go]: I remove the getNodeSummary using 10255 here and change it to use `e2edebug.GetStatsSummary`.

**Comment by liggitt** [comment]
kubelet has different defaults when starting from a config file and from command-line flags... CLI flags came first and defaulted the readonly port to 10250. kubelet file-based config came next and we used that opportunity to improve defaults (see https://github.com/kubernetes/kubernetes/pull/59666)

I still think this is a reasonable step, but it is not a bug.

All the test changes need review/approve from kubelet owners to make sure the surface area is still tested

**Comment by hailkomputer** [comment]
Hello @pacoxu, @bart0sh, v1.28 Bug Triage Shadow here checking in. It is still early in the release cycle, but is this still on track for v1.28 release?

**Comment by pacoxu** [comment]
I prefer to get some reviewers or approvers taking a look in 1.28

**Comment by hailkomputer** [comment]
Hello! Bug triage team shadow here. I want to check the status. The code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about three weeks from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.28, is it still planned for this release?

**Comment by hailkomputer** [comment]
Hi! Bug triage team shadow here.

We are currently in Week 9 of the v1.28 release cycle, and the code freeze is happening next week (Week 10) at [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496)

Please ensure all necessary contributions are made before the deadline if the PR still targets v1.28.

Let me know if you need any assistance.

**Comment by pacoxu** [comment]
> Hi! Bug triage team shadow here.
> 
> We are currently in Week 9 of the v1.28 release cycle, and the code freeze is happening next week (Week 10) at [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496)
> 
> Please ensure all necessary contributions are made before the deadline if the PR still targets v1.28.
> 
> Let me know if you need any assistance.

rebased. Waiting for another review.

**Pull request closed without merging**
